tools: harden scripts

sdavids opened the pull request:
Refs: https://developer.apple.com/library/archive/documentation/OpenSource/Conceptual/ShellScripting/ShellScriptSecurity/ShellScriptSecurity.html#//apple_ref/doc/uid/TP40004268-CH8-SW2
Fix: #57660

sdavids commented:
Here is how one can modify the PATH used by GUI apps—in this case the installer:

```console
sudo launchctl config user --help
Usage: launchctl config <system|user> <parameter> <value>
When given the "system" argument, modifies the configuration for the system
domain. When given the "user" argument on supported platforms, modifies the
configuration for all user domains. You must reboot for changes to take effect.
Note that if a service specifies a conflicting configuration, the service's
parameter will be preferred.

Supported configuration parameters are:

umask <integer as octal>
Modifies the umask(2) applied to services launched in the domain.

path <string>
Modifies the PATH environment variable set on each service in the
domain.
```

aduh95 commented:
Is this file still used? According to https://github.com/nodejs/node/issues/4635#issuecomment-170956761, it's not run by the installer, and the only reference to it I could fine is https://github.com/nodejs/node-convergence-archive/blob/e11fe0c2777561827cdb7207d46b0917ef3c42a7/tools/osx-pkg.pmdoc/02npm.xml 🤔

sdavids commented:
Yes, see 
https://github.com/nodejs/node/issues/57548#issuecomment-2752303400

ljharb commented:
What if the binaries aren't available on that PATH? (they might be brew-installed, for example) It seems correct to me to use whatever is on the PATH.

It's not an attack if it requires you to attack yourself.

sdavids commented:
[System Integrity Protection on your Mac](https://support.apple.com/en-us/102149)

> System Integrity Protection includes protection for these parts of the system:
> […]
> /bin
> […]
> Paths and apps that third-party apps and installers can continue to write to include:
> […]
> /usr/local
> […]
>System Integrity Protection is designed to allow modification of these protected parts only by processes that are signed by Apple and have special entitlements to write to system files, such as Apple software updates and Apple installers.

The *NodeJS* installer extracts to `/usr/local/bin` and creates symlinks in the same directory—this is fine with SIP.
## homebrew
https://docs.brew.sh/FAQ#why-should-i-install-homebrew-in-the-default-location

> The default prefix is:
> […]
/opt/homebrew for macOS on Apple Silicon,
/usr/local for macOS on Intel, and
> […]
Do yourself a favour and install to the default prefix so that you can use our pre-built binary packages. Pick another prefix at your peril!

https://formulae.brew.sh/formula/coreutils

> Commands also provided by macOS and the commands dir, dircolors, vdir have been installed with the prefix "g".
> If you need to use these commands with their normal names, you can add a "gnubin" directory to your PATH with:
> PATH="$HOMEBREW_PREFIX/opt/coreutils/libexec/gnubin:$PATH"

**Note**: This only works for commands started via a terminal app. If you wanted them to be accessible to user apps (like the *NodeJS* installer) you would need to do (on an ARM machine):

```console
$ sudo launchctl config user path '/opt/homebrew/opt/coreutils/libexec/gnubin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin'
$ sudo reboot
```
### MacPorts
https://guide.macports.org/chunked/internals.configuration-files.html

> prefix
> Sets the directory where ports are installed.
> […]
> Default: /opt/local


https://trac.macports.org/wiki/FAQ#defaultprefix

> Traditionally […] /usr/local. However, having MacPorts under /usr/local would be error-prone for precisely that reason. Many other software packages and packaging systems install into /usr/local, and could accidentally overwrite what MacPorts has installed, or vice versa […] MacPorts doesn't want to be a victim of that, and /opt/local provides the splendid isolation […]

## TL;DR

When [SIP](https://support.apple.com/en-us/102149) is enabled (default in all modern macOS versions) then third-party installers, i.e. the *NodeJS* installer, and superusers cannot modify the contents of `/bin`.

That is why this hardening is useful.

A malicious entity cannot prepend the user/system path on a user’s host via `sudo launchctl config [user|systen] path` and then wait for someone to run the *NodeJS* installer.

The malicious `ln` would then run with administrator privileges and 💥

ljharb commented:
so the attack vector is that i can inject an unprivileged executable, and by way of another script, be sudo'd into a privileged context? 🤔

sdavids commented:
_macOS_ is a multi-user system, even if most people do not use it that way.

---

The attack vector is:

1. prepending the user path with a directory (e.g. homebrew or MacPorts bin dir); this could have been done by your company's admins for various reasons
2. placing a malicious `ln` into that directory
3. executing the current _NodeJS_ installer

Quoting Apple here:

https://developer.apple.com/library/archive/documentation/OpenSource/Conceptual/ShellScripting/ShellScriptSecurity/ShellScriptSecurity.html#//apple_ref/doc/uid/TP40004268-CH8-SW2

> Always specify absolute or relative paths when executing binaries or other scripts.

sdavids commented:
> This seems to indicate it is _not_ used, AFAICT it only mentions

The scripts are executed when you execute the installer, not when you use `pkgutil --expand`.

I just found out that there is another _postinstall_ script in the repo:

https://github.com/nodejs/node/blob/main/tools/macos-installer/pkgbuild/npm/scripts/postinstall

aduh95 commented:
I've opened #57667 to remove it

sdavids commented:
I found where the scripts are used:


https://github.com/npm/node/blob/2bf0bd2c9d25b810076eec6b2fdaf333f4a1eac2/Makefile#L179-L1183

```
	pkgbuild --version $(NPMVERSION) \
		--identifier org.nodejs.npm.pkg \
		--root $(MACOSOUTDIR)/dist/npm \
		--scripts ./tools/macos-installer/pkgbuild/npm/scripts \
			$(MACOSOUTDIR)/pkgs/npm-$(NPMVERSION).pkg
```

```
$ man pkgbuild
NAME
     pkgbuild – Build a macOS Installer component package from on-disk files
...
     --scripts scripts-path
             […] If this directory contains scripts named preinstall and/or postinstall, these will be ran as the top-level scripts of the package […]
```

sdavids commented:
I have updated the PR.

I moved the symlink creation into the _Makefile_ thereby making the _postinstall_ script obsolete.

This will fix #57660 and #57548.

aduh95 commented on Makefile:
```diff
@@ -1176,6 +1176,9 @@ endif
 	pkgbuild --version $(FULLVERSION) \
 		--identifier org.nodejs.node.pkg \
 		--root $(MACOSOUTDIR)/dist/node $(MACOSOUTDIR)/pkgs/node-$(FULLVERSION).pkg
+	mkdir -p $(MACOSOUTDIR)/dist/npm/usr/local/bin
+	ln -sf ../lib/node_modules/npm/bin/npm-cli.js $(MACOSOUTDIR)/dist/npm/usr/local/bin/npm
+	ln -sf ../lib/node_modules/npm/bin/npx-cli.js $(MACOSOUTDIR)/dist/npm/usr/local/bin/npx
```
That doesn't seem correct, we don't want to install those on the release machine
```suggestion
```

sdavids replied:
You can either place these files on the release machine before invoking the `pkgbuild` and then the are inside and no postinstall is needed or you need the postinstall script.

One could remove them right after the `pkgbuild` though.

sdavids replied:
I am also unsure what you mean by "release machine".

These symlinks are created inside the build folder `out` which is gitignored.

aduh95 replied:
Moving the `unlink` call to after invocating `pkgbuild` sounds like a better solution

sdavids replied:
```shell
unlink $(MACOSOUTDIR)/dist/node/usr/local/bin/npx
```

The `unlink`s are for the `npm` and `npx` symlinks in the `node` and not the `npm` directory; these would have been placed in the `org.nodejs.node.pkg` package. 

The symlinks created by this PR will be placed in the `org.nodejs.npm.pkg` package.

One could add 2 additional `unlink` statements for `npm` and `npx` after the `pkgbuild` though.

But I am not sure if it is worth it because aren't these links on an ephemeral directory on a build machine?

sdavids commented:
Make stalebot happy

sdavids commented:
Still relevant

aduh95 commented:
/cc @nodejs/platform-macos @ryanaslett

sdavids commented:
Stalebot…
